### PR TITLE
fix(mirror): fix issues with limits on transaction speed

### PR DESCRIPTION
**Pull request by marcelo-gonzalez**
There has been an issue where the rate of transactions sent can only go so high, and when we try to raise it with the "tx_batch_interval" option, it hits a bottleneck and fails to send transactions fast enough. What is happening is that the `main_loop()` function handles many other things than sending transactions, and shares a thread with the indexer for the target chain. The biggest bottleneck is the fact that building new sets of transactions to send in `queue_txs()` after each batch of transactions we send actually can take a long time, so that the early break in that function when we're 20 milliseconds away from needing to send more transactions is not enough. Then also sometimes we are held up by the indexer sending a new block and us processing it.

So to fix this, quite a lot needs to be refactored so that we can put these components in different threads. We pass around a `Mutex<TxTracker>`, and modify `on_target_block()` so that it's not async, because otherwise we wouldn't be able to lock the mutex and call that function in a new thread, since the resulting future would not be `Send`. The main thread handles queueing new blocks worth of transactions, and calling all the async `TxTracker` functions, and we start a new thread to run the loop that sends transactions, and another to run the indexer and call `TxTracker::on_target_block()`

### CODE REVIEW

**Comment by VanBarbascu** [comment]
@marcelo-gonzalez, awesome job refactoring this! 
There are a few clippy errors that need to be addressed before merging.

File: tools/mirror/src/lib.rs [hunk 1]
```diff
@@ -1747,13 +1961,55 @@ impl<T: ChainAccess> TxMirror<T> {
...
+        let (clients_tx, clients_rx) = tokio::sync::oneshot::channel();
+        let (target_indexer_done_tx, target_indexer_done_rx) =
+            tokio::sync::oneshot::channel::<anyhow::Result<()>>();
+        let (unstake_tx, unstake_rx) = mpsc::channel(10);
+
+        let db = self.db.clone();
+        let target_height2 = target_height.clone();
+        let target_head2 = target_head.clone();
+        let tracker2 = tracker.clone();
+        let arbiter = actix::Arbiter::new();
```
Comment (VanBarbascu) [inline tools/mirror/src/lib.rs:1981]: Can you use more descriptive names here?

**Review by VanBarbascu — APPROVED** [review]
Overall, the logic appears solid to me. Nice work!

**Comment by VanBarbascu** [comment]
It is also worth adding a small diagram of the system to show how the data flows.

**Comment by VanBarbascu** [comment]
One thing that I missed is what will happen to the tx that are buffered in the TxTracker if we restart the node?

**Comment by marcelo-gonzalez** [comment]
> One thing that I missed is what will happen to the tx that are buffered in the TxTracker if we restart the node?

when we call [`set_last_source_height`](https://github.com/near/nearcore/blob/ec2ece32482c23443f85cd50f999914e717a50a6/tools/mirror/src/lib.rs#L213), it saves the last height we sent txs for on disk. then next time we start up, we start building and sending transactions [from that point](https://github.com/near/nearcore/blob/ec2ece32482c23443f85cd50f999914e717a50a6/tools/mirror/src/lib.rs#L1724)

**Comment by marcelo-gonzalez** [comment]
> It is also worth adding a small diagram of the system to show how the data flows.

hm ok good suggestion, maybe we can do that in another PR

File: tools/mirror/src/lib.rs (hunk 1 above)
Comment (marcelo-gonzalez) [reply tools/mirror/src/lib.rs:1981]: Are you referring to the naming pattern `let X2 = X.clone()`, or the names `X` themselves? Do you have suggestions for what you had in mind? Like `target_head` referring to the head hash of the target chain already seems kind of descriptive to me, no? If you have suggestions, I'm happy to hear them though

Comment (marcelo-gonzalez) [reply tools/mirror/src/lib.rs]: renamed the arbiters, in case thats what you meant

**Comment by marcelo-gonzalez** [comment]
> There are a few clippy errors that need to be addressed before merging.

Ok i fixed these now, it was not just a simple fix actually. had to unwind lots of mutexes used across awaits

**Comment by VanBarbascu** [comment]
> > It is also worth adding a small diagram of the system to show how the data flows.
> 
> hm ok good suggestion, maybe we can do that in another PR

Thanks! Please don't forget about this

**Pull request merged**